### PR TITLE
gh-89727: Fix pathlib.Path.walk RecursionError on deep trees

### PR DESCRIPTION
**Pull request by zmievsa**
Use a stack to implement pathlib.Path.walk iteratively instead of recursively to avoid hitting recursion limits on deeply nested trees.

* Issue: gh-89727

### CODE REVIEW

**Comment by Ovsyanka83** [comment]
TODO:
* Add a news blip
* Add a test that checks for deeply nested directories
* Figure out a better name for is_result

File: Lib/pathlib.py
Comment (kalvdans) [inline Lib/pathlib.py]: Unused import.

File: Lib/test/test_pathlib.py
Comment (carljm) [inline Lib/test/test_pathlib.py]: Now that it was merged in the `os.walk` PR, maybe use `test.support.set_recursion_limit` here? Although in practice the functionality is the same, the name `infinite_recursion` is inappropriate to this use and may confuse the reader regarding the purpose of the test (this test has nothing to do with infinite recursion.)
```suggestion
        with set_recursion_limit(40):
```

Comment (zmievsa) [reply Lib/test/test_pathlib.py]: Thank  you! That was my intention all along to also use `set_recursion_limit` but I wanted to wait until it gets merged to prevent duplicate commits and possible merge conflicts that could arise if I did anything slightly differently.

**Review by carljm — APPROVED** [review]
LGTM. I assume because of the initial underscore we don't have to worry about people who may have subclassed `Path` and overridden the `_walk` method (now to no effect.)

**Comment by Ovsyanka83** [comment]
@carljm remember that Path.walk has only been created in 3.12 which means that 99% of libraries/projects do not and cannot depend on it yet.

**Comment by Ovsyanka83** [comment]
@barneygale   any estimates of when you will have time to take a look at it? No pressure or rush though.

**Comment by barneygale** [comment]
I'll look at this within the next few days!

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]: `os.scandir()` gives us names in an arbitrary order. Should we bother reversing them here?

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]:
```suggestion
            is_path, arg = stack.pop()
            if is_path:
                path = arg
            else:
                yield arg
                continue
```

Using the variable `top` to mean 'top of the walk stack' when we already have have a function argument named `top_down` referring to the 'top of the directory tree' could be misleading. 

Perhaps `arg` would be better, and swap the meaning of the boolean? Not too sure about this.

**Review by barneygale — CHANGES_REQUESTED** [review]
This looks like it will chomp through all my system memory if it encounters a symlink loop. Is that the case? If so, we might want to either detect loops and bail out, or add a *max_depth* argument with a reasonable default.

Edit: on Linux, there seems to be a limit of ~40 symlinks in the path, so I don't get any issues. I'll check on Mac shortly.

File: Lib/pathlib.py
Comment (zmievsa) [reply Lib/pathlib.py]: It seems to me that it's only arbitrary in the sense of it not having any clear rules about sorting. However, its order **is** stable which is why un-reversing them will make a few of the tests fail. And we do not want any tests to fail to keep as much compatibility with `os.walk`'s ordering rules as possible.

File: Lib/pathlib.py
Comment (zmievsa) [reply Lib/pathlib.py]: I am a bit confused. top_down is not the top of the directory tree. It is the boolean flag that indicates whether we start from top and go down or not.

Or did I misunderstand what you meant?

Comment (barneygale) [reply Lib/pathlib.py]: Sorry for not being clear. I mean that the meaning _of 'top'_ in `top_down` is 'the top of the directory tree'.

File: Lib/pathlib.py
Comment (barneygale) [reply Lib/pathlib.py]: It might be good to get a core dev opinion on this.

One thing that only just occurred to me: users of `top_down=True` might expect us to visit directories in the same order they appear in the yielded `dirnames`, which would make the `reversed()` call worthwhile. Or maybe not, as it's not documented anywhere.

**Comment by barneygale** [comment]
Idea: make the stack is a list of path objects and 3-tuples, _without_ a leading boolean to differentiate them. I think this could improve code clarity and _might_ provide a speed-up.

```python
path = stack.pop()
if isinstance(path, tuple):  # might be faster with a 'if not top_down and' in front.
    yield path
    continue
```

You could simplify the stack additions a bit:

```python
stack.extend(path._make_child_relpath(dirname) for dirname in dirnames)
```

**Comment by Ovsyanka83** [comment]
@barneygale I compared your solution with my previous one and got a significant speedup:

image


image

Thanks!

File: Lib/pathlib.py
Comment (zmievsa) [reply Lib/pathlib.py]: Fixed!

File: Lib/pathlib.py
Comment (zmievsa) [reply Lib/pathlib.py]: 1. I believe that users will expect Path.walk ordering to be consistent with the ordering of the recursive `os.walk` that existed for 20+ years. And I think our tests verify that it is consistent now but won't be consistent if we remove `reverse`.
2. We only mention scandir in the error handling section in `walk`'s docs so I think the absolute majority of users do not even think about scandir being used within walk, and definitely do not know much about its ordering rules. 

Sorry if I'm repeating the same ideas over and over. I'm totally fine with waiting for core dev's opinion on this. Though we should ping someone specific.

Comment (barneygale) [reply Lib/pathlib.py]: @AlexWaygood what do you think? The `reversed()` call is there to ensure we visit subdirectories in the same order as before this patch, but we don't document that order, and in fact it's arbitrary (but stable across runs) as we're using `os.scandir()` under-the-hood. Should it be removed for simplicity's sake? (+ tiny performance improvement, I assume)

Comment (barneygale) [reply Lib/pathlib.py]: If we decide to remove the `reversed()` call then the tests will need adjustment. If possible I'd love for them to stop making assertions about the order of items in `dirnames` and `filenames`, and stop making assertions about the order in which sibling directories are visited. Is that doable @Ovsyanka83, or is it a bit of a horror show?

Comment (zmievsa) [reply Lib/pathlib.py]: I do not think like it is a big problem. Many tests will need to be reworked and some thought will need to be put into making sure our tests stay effective but it is definitely doable.

Though I would prefer to stick with the `reversed()` version so let's hear @AlexWaygood 's opinion on this to help us decide.

Comment (AlexWaygood) [reply Lib/pathlib.py]: I don't _think_ the `reversed()` call should have a noticeable impact on performance, due to the fact that `reversed()`, unlike `sorted()`, is a _lazy_ iterator: it uses a specialised iterator to iterate over lists backwards, rather than creating a new list in memory:

```pycon
>>> reversed([])
<list_reverseiterator object at 0x00000299ED6A2CB0>
```

(Having said that, if you've got numbers to prove me wrong, I'd be very interested to see them ;)

I also agree with @Ovsyanka83 that, while `Path.walk()` is a new function (so we're not bound by backwards-compatibility constraints), it certainly seems like it would probably be the least surprising thing to match the iteration order that users have grown accustomed to with `os.walk()`.

Comment (barneygale) [reply Lib/pathlib.py]: Makes sense to me, thank you both.

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]:
```suggestion
            stack.extend(path._make_child_relpath(dirname) 
                         for dirname in reversed(dirnames))
```

Or perhaps

```suggestion
            stack += [path._make_child_relpath(dirname) 
                      for dirname in reversed(dirnames)]
```

Might be faster, and seems no less clear.

Comment (zmievsa) [reply Lib/pathlib.py]: In my experience, listcomp is faster in such cases but the primary benefit of this is how clear it is.

**Review by barneygale — APPROVED** [review]
Looks good to me!

**Comment by Ovsyanka83** [comment]
@AlexWaygood what would be the next steps for this PR? I am ready to continue improving it

**Comment by AlexWaygood** [comment]
Sorry @Ovsyanka83, I'll try to take a proper look soon!

File: Lib/pathlib.py
Comment (brettcannon) [inline Lib/pathlib.py]:
```suggestion
            # We may not have read permission for self, in which case we can't
            # get a list of the files the directory contains. os.walk()
            # always suppressed the exception in that instance, rather than
            # blow up for a minor reason when (say) a thousand readable
            # directories are still left to visit. That logic is copied here.
```

File: Lib/test/test_pathlib.py
Comment (brettcannon) [inline Lib/test/test_pathlib.py]:
```suggestion
        recursion_limit = 40
        # directory_depth > recursion_limit
        directory_depth = recursion_limit + 10
        base = pathlib.Path(os_helper.TESTFN, 'deep')
        path = pathlib.Path(base, *(['d'] * directory_depth))
        path.mkdir(parents=True)

        with set_recursion_limit(recursion_limit):
```

**Review by brettcannon — APPROVED** [review]

**Comment by barneygale** [comment]
@AlexWaygood gentle nudge to review this when you have the time!

**Review by CuriousLearner — APPROVED** [review]
LGTM.

Thanks for your contribution.

**Comment by barneygale** [comment]
@AlexWaygood https://youtu.be/CdqoNKCCt7A

File: Lib/pathlib.py
Comment (barneygale) [inline Lib/pathlib.py]: Sorry for the late feedback. It might be better to name this `paths` rather than `stack`. In the future we might add a breadth-first walk, in which case this variable would be treated as a queue rather than a stack.

(also just generally good practice not to name things merely for their data structure)

Comment (zmievsa) [reply Lib/pathlib.py]: On it!

Comment (zmievsa) [reply Lib/pathlib.py]: Fixed. Ready to resolve.

Comment (barneygale) [reply Lib/pathlib.py]: Tyvm!

**Review by barneygale — APPROVED** [review]

**Comment by AlexWaygood** [comment]
We made him a core dev! https://discuss.python.org/t/vote-to-promote-barney-gale/24801

**Comment by barneygale** [comment]
@Ovsyanka83 thanks for your patience and for working on this, it's a neat patch :). I'm hoping to build an iterative version of `glob()` atop!

**Pull request merged**
